MarkupsToModel/MRML/vtkMRMLMarkupsToModelNode.cxx

```diff
@@ -267,165 +248,318 @@ void vtkMRMLMarkupsToModelNode::SetAndObserveMarkupsNodeID( const char* markupsI
 }
 
-std::string vtkMRMLMarkupsToModelNode::GetDisplayNodeName()
+void vtkMRMLMarkupsToModelNode::SetOutputOpacity( double outputOpacity )
 {
-  if(this->ModelNode == NULL)
+  vtkMRMLModelNode* modelNode = vtkMRMLModelNode::SafeDownCast( this->GetNodeReference( OUTPUT_MODEL_ROLE ) );
+  if ( modelNode == NULL )
   {
-    return std::string(this->GetID()).append("Display");
+    return;
   }
-  else
+  
+  vtkMRMLModelDisplayNode* displayNode = vtkMRMLModelDisplayNode::SafeDownCast( modelNode->GetDisplayNode() );
+  if ( displayNode == NULL )
   {
-    return std::string(this->ModelNode->GetDisplayNode()->GetName());
+    this->createAndObserveModelDisplayNode();
+    displayNode = vtkMRMLModelDisplayNode::SafeDownCast( modelNode->GetDisplayNode() );
   }
+
+  displayNode->SetOpacity( outputOpacity );
 }
 
-void vtkMRMLMarkupsToModelNode::SetOutputOpacity(double outputOpacity)
+void vtkMRMLMarkupsToModelNode::SetOutputVisibility( bool outputVisibility )
 {
-  if(this->ModelNode != NULL)
+  vtkMRMLModelNode* modelNode = vtkMRMLModelNode::SafeDownCast( this->GetNodeReference( OUTPUT_MODEL_ROLE ) );
+  if ( modelNode == NULL )
+  {
+    return;
+  }
+  
+  vtkMRMLModelDisplayNode* displayNode = vtkMRMLModelDisplayNode::SafeDownCast( modelNode->GetDisplayNode() );
+  if ( displayNode == NULL )
+  {
+    this->createAndObserveModelDisplayNode();
+    displayNode = vtkMRMLModelDisplayNode::SafeDownCast( modelNode->GetDisplayNode() );
+  }
+
+  if ( outputVisibility )
+  {
+    displayNode->VisibilityOn();
+  }
+  else
   {
-    //this->ModelNode->GetModelDisplayNode->SetOpacity(outputOpacity);
-    this->ModelNode->GetDisplayNode()->SetOpacity(outputOpacity);
+    displayNode->VisibilityOff();
   }
 }
 
-void vtkMRMLMarkupsToModelNode::SetOutputVisibility(bool outputVisibility)
+void vtkMRMLMarkupsToModelNode::SetOutputIntersectionVisibility( bool outputIntersectionVisibility )
 {
-  if(this->ModelNode != NULL)
+  vtkMRMLModelNode* modelNode = vtkMRMLModelNode::SafeDownCast( this->GetNodeReference( OUTPUT_MODEL_ROLE ) );
+  if ( modelNode == NULL )
   {
-    if(outputVisibility)
-    {
-      this->ModelNode->GetDisplayNode()->VisibilityOn();
-    }
-    else
-    {
-      this->ModelNode->GetDisplayNode()->VisibilityOff();
-    }
+    return;
+  }
+  
+  vtkMRMLModelDisplayNode* displayNode = vtkMRMLModelDisplayNode::SafeDownCast( modelNode->GetDisplayNode() );
+  if ( displayNode == NULL )
+  {
+    this->createAndObserveModelDisplayNode();
+    displayNode = vtkMRMLModelDisplayNode::SafeDownCast( modelNode->GetDisplayNode() );
+  }
+
+  if ( outputIntersectionVisibility )
+  {
+    displayNode->SliceIntersectionVisibilityOn();
+  }
+  else
+  {
+    displayNode->SliceIntersectionVisibilityOff();
   }
 }
 
-void vtkMRMLMarkupsToModelNode::SetOutputIntersectionVisibility(bool outputIntersectionVisibility)
+void vtkMRMLMarkupsToModelNode::SetOutputColor( double redComponent, double greenComponent, double blueComponent )
```

Comment:
All these Set... functions that manipulate the display options should be moved out of this class.